src/kudu/consensus/raft_consensus.cc

```diff
@@ -3037,6 +3035,33 @@ Status RaftConsensus::SetCommittedConfigUnlocked(const RaftConfigPB& config_to_c
   return Status::OK();
 }
 
+void RaftConsensus::ScheduleTermAdvancementCallback(int64_t new_term) {
+  WARN_NOT_OK(
+      raft_pool_token_->SubmitFunc(
+        std::bind(&RaftConsensus::DoTermAdvancmentCallback,
+                  shared_from_this(),
+                  new_term)),
+      LogPrefixThreadSafe() + "Unable to run term advancement callback");
+}
+
+void RaftConsensus::DoTermAdvancmentCallback(int64_t new_term) {
+  // Simply execute the registered callback for term advancement.
+  if (tacb_) tacb_(new_term);
+}
+
+void RaftConsensus::ScheduleNoOpReceivedCallback(const ReplicateRefPtr& msg) {
+  WARN_NOT_OK(
+      raft_pool_token_->SubmitFunc(
+        std::bind(&RaftConsensus::DoNoOpReceivedCallback,
+                  shared_from_this(),
+                  msg)),
```

Comment:
Since msg is a reference to ReplicateRefPtr, what is the guarantee that it is still valid when the callback gets scheduled?

Reply:
It's a ref counted pointer defined in `ref_counted_replicate.h` so as long as a valid reference is present in any scope it should not get deallocated.